script/model/forecast.py

```diff
@@ -0,0 +1,520 @@
+
+    rates = tpcc_rates * pattern_iter
+    gen_forecast_data(test_case.xml_config, rates)
+
+    # Run the actual test
+    ret_val, _, stderr = run_command(test_case.test_command,
+                                     test_case.test_error_msg,
+                                     cwd=test_case.test_command_cwd)
+    if ret_val != ErrorCode.SUCCESS:
+        LOG.error(stderr)
+        return False
+
+    # Clean up, disconnect the DB
+    db_server.stop_db()
+    db_server.delete_wal()
+
+    if not Path(DEFAULT_QUERY_TRACE_FILE).exists():
+        LOG.error(
+            f"Missing {DEFAULT_QUERY_TRACE_FILE} at CWD after running OLTP TPCC")
+        return False
+
+    return True
+
+
+class DataLoader:
+    # Hardcoded timestamp column index in the query_trace file
+    TS_IDX = 1
+
+    def __init__(
+            self,
+            query_trace_file: str = DEFAULT_QUERY_TRACE_FILE,
+            test: bool = False,
+            eval_size: int = EVAL_DATA_SIZE,
+            seq_len: int = SEQ_LEN,
+            horizon_len: int = HORIZON_LEN) -> None:
+        """
+        :param query_trace_file: Where to process the query trace
+        :param test: If the Loader is for testing
+        :param eval_size:  Number of data points used for evaluation(testing)
+        :param seq_len: Length of a sequence
+        :param horizon_len: Horizon length
+        """
+        self.query_trace_file = query_trace_file
+
+        self.seq_len = seq_len
+        self.horizon_len = horizon_len
+        self.scaler = MinMaxScaler(feature_range=(-1, 1))
+        self.test_mode = test
+        self.eval_data_size = eval_size
+
+        self._load_data()
+
+    def _load_data(self) -> None:
+        LOG.info(f"Loading data from {self.query_trace_file}")
+        # Load data from the files
+        with open(self.query_trace_file, newline='') as csvfile:
+            reader = csv.DictReader(csvfile)
+            data = np.array(
+                [[int(r['query_id']), int(r[' timestamp'])] for r in reader])
+
+            data = self._aggregate_data(data, INTERVAL_MS)
+            self.train_raw_data, self.test_raw_data = self._split_data(data)
+            LOG.info(
+                f"Loaded data successfully: num_test={len(self.test_raw_data)}, num_train={len(self.train_raw_data)}, "
+                f"seq_len={self.seq_len}, horizon={self.horizon_len}")
+
+    def _normalize_data(self, data: np.ndarray) -> np.ndarray:
+        norm_data = self.scaler.fit_transform(data.reshape(-1, 1))
+        return norm_data
+
+    def _aggregate_data(self, data: np.ndarray,
+                        interval_ms: int) -> Optional[np.ndarray]:
+        """
+        Aggregate multiple time-series data points into one bucket to generate a new time-series
+        data set.
+        It assumes the original time-series data has a Timestamp value, and the generated time-series
+        will be the number of data points fall into an interval for each interval in the original dataset
```

Comment:
Okay, so what we need is slightly different than this. Instead of aggregating all queries together right now, we should generate one time-series for each query (`query_id`). It's only when doing the training and inference we're going to further aggregate all the time-series together (assuming that all queries belong to the same cluster). But after the model inference, we need to split the single time-series prediction back to the predictions for each query, based on the ratio between the total occurrences of these queries in the training data. In other words, even though the model only predicts one time-series, we should generate the same number of time-series predictions as the number of queries at the end. We probably should have a `Forecaster` class that handles this logic. Is this explanation clear?

Reply:
So something like this: 
A query trace file => multiple time-series of certain interval for each query id => aggregate multiple time-series into one time-series for each cluster and note down the ratio => for each cluster, a model will produce a prediction => convert the prediction into multiple time-series for different query id according to the ratio. 

And the ratio would just be the total number of events of a single query id over all events in the cluster? (Assuming queries in the same cluster would have the same pattern)

Reply:
Yes! Your summary is exactly right!